File: ruby/04-switch-audio.rb
require 'rpi_gpio'

# Get filename from command line arguments
filename = ARGV[0]
raise "Incorrect number of arguments. Usage: ruby 04-switch-audio.rb song.mp3" unless filename

# Run `gpio readall` to confirm BCM pin number(s)
pin_in = 18   # button

RPi::GPIO.set_numbering(:bcm)
RPi::GPIO.set_warnings(0)
RPi::GPIO.setup(pin_in, as: :input, pull: :down)

def play(filename)
  # Use commandline omxplayer to play file:
  # http://www.raspberrypi.org/documentation/usage/audio/
  system("omxplayer '#{filename}'")
end

begin
  loop do
    if RPi::GPIO.high?(pin_in) # button input state is "on"
      play(filename)
      sleep(0.2)
    end
  end
rescue Interrupt
  puts 'Done'
ensure
  RPi::GPIO.clean_up
end

File: node/04-switch-audio.js
var Gpio     = require('onoff').Gpio,
    fs       = require('fs'),
    lame     = require('lame'),
    Speaker  = require('speaker'),
    button   = new Gpio(18, 'in', 'both'),	// Run `gpio readall` to confirm BCM pin number(s)
    filename = process.argv[2];				// Get filename from command line arguments

if (!filename) {
	console.log('Incorrect number of arguments. Usage: node 04-switch-audio.rb song.mp3');
	exit();
}

// Watch for button state to change from "0" to "1"
button.watch(function(err, value) {
	if (err) exit();
	if (value == 1) {
		console.log('Button pressed');
		// Play audio file
		fs.createReadStream(filename)
			.pipe(new lame.Decoder())
			.on('format', function (format) {
				this.pipe(new Speaker(format));
			});
	}
});

function exit() {
	// Cleanup
	button.unexport();
	process.exit();
}

// Exit on kill signal (Ctrl+C)
process.on('SIGINT', exit);